Added function to write timezone name to /etc/timezone if necessary. …

jbrockmendel opened the pull request:
… This is necessary on Raspbian.

cdown commented:
Hey there! Thanks for the code. A few general comments:

- Tests are broken, you can see this on the Travis CI build above. You'll need to mock out appropriately before this can be merged.
- Please also add tests for this case.

I'll also add some inline comments.

cdown commented:
Awesome -- thank you! :-)

jbrockmendel commented:
Addressed most of the issues you brought up.  

(Not sure whether to edit this pull request or open a new one)

The race condition I'm not sure how to handle-- or even if it needs handling.  If the file exists at time A and is deleted at time A+epsilon, does that really matter?  We're about to overwrite it anyway.  Actually asking, because I don't know what best-practices are here.

Added unit tests for cases where the file does or does not exist, no special cases for user-supplied paths.  

For reasons I haven't yet figured out, the e2etests do not run locally with "tox".  Still need to add tests for permissions issues. (Is it decorous to wait to send this stuff until the rough edges are sorted out, or to do smaller bits?)

cdown commented on tzupdate.py:
```diff
@@ -118,9 +153,9 @@ def parse_args(argv):
     )
     parser.add_argument(
-        "-z", "--zoneinfo-path",
+        '-z', '--zoneinfo-path',
```
[blocker] Don't change unrelated style stuff in a diff, it clouds `git blame`

cdown commented on tzupdate.py:
```diff
@@ -102,6 +102,41 @@ def link_localtime(timezone, zoneinfo_path, localtime_path):
+        try:
+            with open(etc_timezone_path, 'w') as fd:
+                fd.write(timezone+'\n')
+        except (IOError,OSError) as thrown_exc:
```
[style] There should be space between the two tuple elements.

cdown commented on tests/e2e_tests.py:
```diff
@@ -39,20 +46,28 @@ def test_explicit_paths(link_localtime_mock, get_timezone_for_ip_mock):
     )
+    assert_true(
+        export_etc_timezone_mock.called_once_with(
+            FAKE_TIMEZONE, etc_timezone_path,
```
[query] Are you sure this works? It looks like `etc_timezone_path` was not added to the arg list.

jbrockmendel replied:
I'm generally *unsure* of most of what's going on in e2e_tests, as it still doesn't get run by tox locally.

That said, etc_timezone_path is defined in line 40.

cdown replied:
Are you running with `tox -e quick`? It runs E2E tests fine for me.

jbrockmendel replied:
No, just "tox".  It says it ran 19 tests for each of 2.7, 3.4, 3.5.  That suggests that it is in fact running everything in both files.

I think I've figured it out.  The issue is why it is raising PermissionErrors within Travis but not locally.  I think it is because I am on OSX, which does not have the /etc/timezone file, so the function is getting called but not reaching the point where it would hit the PermissionError.

cdown commented on tests/e2e_tests.py:
```diff
@@ -39,20 +46,28 @@ def test_explicit_paths(link_localtime_mock, get_timezone_for_ip_mock):
     args = ['-a', ip_addr]
     tzupdate.main(args)
     assert_true(get_timezone_for_ip_mock.called_once_with(ip_addr))
+    assert_true(export_etc_timezone_mock.called)
```
[optional] This isn't essential to this test, I'd leave it out.

jbrockmendel replied:
If you look at the Travis logs from build 76, you'll see that without this call an error is raised.  The traceback ends with:

  File "/home/travis/build/cdown/tzupdate/tzupdate.py", line 133, in export_etc_timezone
    raise thrown_exc
IOError: [Errno 13] Permission denied: '/etc/timezone'

cdown replied:
The assert_true has no bearing on that happening or not, only the mock does.

That is, you can add the mock, you don't have to then assert `.called` on it.

cdown commented on tests/e2e_tests.py:
```diff
@@ -62,3 +77,9 @@ def test_explicit_timezone(link_localtime_mock):
             tzupdate.DEFAULT_ZONEINFO_PATH, tzupdate.DEFAULT_LOCALTIME_PATH,
         )
     )
+    assert_true(
```
[optional] Use `assert_called_one_with` for a decent failure message

jbrockmendel replied:
Sure.  I was trying to mimic the style of the assertion immediately before this one.  What's the pertinent distinction?

cdown commented on tests/unit_tests.py:
```diff
@@ -141,3 +141,34 @@ def test_link_localtime_localtime_missing_no_raise(symlink_mock, isfile_mock,
+                                    )
+
+    eq(wrote_etc_timezone, False)
+
```
[optional] I'd also check that open() wasn't called.

jbrockmendel replied:
Good idea.

cdown commented on tzupdate.py:
```diff
@@ -102,6 +102,41 @@ def link_localtime(timezone, zoneinfo_path, localtime_path):
+                    )
+                thrown_exc.message = msg
+                raise thrown_exc
+            elif thrown_exc.errno != errno.ENOENT:
```
[blocker] This is going to silently ignore any other error since you don't reraise under other conditions. Just remove this other elif.

jbrockmendel replied:
I don't understand.  My intention was to add the "are you root?" message before raising in one specific case, and to just raise in all other cases.  Trying to copy the flow/form of link_localtime.  What am I missing?

cdown replied:
You have three cases:

1. EACCES -> raise
2. ENOENT -> raise
3. Anything else -> drop off the end of the if/elif chain and don't raise (bad!)

cdown commented:
Implemented in 27c70e353f9b990d1b9bdacee4e897da1ddd2822.
